File: Traditional-Important-Algorithms/exercises_py/09_rate_limiting/rate09_graceful_degradation.py
# I AM NOT DONE

"""
Exercise: Graceful Degradation Pattern

Graceful Degradation is a fault tolerance pattern that maintains partial
functionality when a system component fails, rather than complete failure.

Strategies:
- Fallback to cached/default values
- Reduced functionality mode
- Alternative service endpoints
- Feature toggling based on system health

Your task: Implement graceful degradation strategies.

Key concepts:
- Service health monitoring
- Fallback mechanisms
- Quality of service levels
- Partial failure handling
"""

import time
from typing import Callable, TypeVar, Dict, List, Optional
from enum import Enum


T = TypeVar('T')


class ServiceHealth(Enum):
    """Service health states"""
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    FAILING = "failing"


class QualityLevel(Enum):
    """Quality of service levels"""
    FULL = "full"        # All features enabled
    REDUCED = "reduced"  # Some features disabled
    MINIMAL = "minimal"  # Only critical features


class ServiceEndpoint:
    """Service endpoint with primary, fallback, and cache"""

    def __init__(self, primary: Callable[[], T],
                 fallback: Optional[Callable[[], T]] = None):
        self.primary = primary
        self.fallback = fallback
        self.cache = None
        self.health = ServiceHealth.HEALTHY
        self.last_success = None
        self.failure_count = 0


class GracefulDegradation:
    """Graceful degradation system"""

    def __init__(self, health_check_interval: float, failure_threshold: int):
        """
        Initialize graceful degradation system

        Args:
            health_check_interval: Time between health checks (seconds)
            failure_threshold: Failures needed to mark as failing
        """
        # TODO: Initialize graceful degradation system
        pass

    def register_service(self, name: str,
                        primary: Callable[[], T],
                        fallback: Optional[Callable[[], T]] = None):
        """Register a service with primary and optional fallback"""
        # TODO: Register a service with primary and optional fallback
        # - Create ServiceEndpoint with given functions
        # - Initialize health as HEALTHY
        # - Set failure_count to 0
        pass

    def call_service(self, name: str):
        """Call service with graceful degradation"""
        # TODO: Call service with graceful degradation
        #
        # Strategy:
        # 1. Get service, raise exception if doesn't exist
        # 2. Try primary service
        # 3. If success:
        #    - Update cache with result
        #    - Reset failure_count
        #    - Update health to HEALTHY
        #    - Update last_success
        #    - Return result
        # 4. If failure:
        #    - Increment failure_count
        #    - Update health based on failure_count
        #    - Try fallback if available
        #    - If fallback succeeds, return result but keep degraded health
        #    - If no fallback or fallback fails, try cache
        #    - If cache exists, return cached value
        #    - Otherwise raise exception
        pass

    def get_health(self, name: str) -> Optional[ServiceHealth]:
        """Return health status of service"""
        # TODO: Return health status of service
        pass

    def get_quality_level(self, name: str) -> Optional[QualityLevel]:
        """Determine quality level based on health"""
        # TODO: Determine quality level based on health
        # - HEALTHY -> FULL
        # - DEGRADED -> REDUCED
        # - FAILING -> MINIMAL
        pass

    def update_cache(self, name: str, value: T):
        """Manually update cache for a service"""
        # TODO: Manually update cache for a service
        pass

    def clear_cache(self, name: str):
        """Clear cache for a service"""
        # TODO: Clear cache for a service
        pass

    def reset_health(self, name: str):
        """Reset service health to HEALTHY"""
        # TODO: Reset service health to HEALTHY
        # - Set health to HEALTHY
        # - Reset failure_count to 0
        pass


class FeatureFlags:
    """Feature flag manager for graceful degradation"""

    def __init__(self):
        """Initialize feature flags"""
        # TODO: Initialize feature flags
        pass

    def register_feature(self, name: str, dependencies: List[str]):
        """Register feature with service dependencies"""
        # TODO: Register feature with service dependencies
        # - Add to flags as enabled by default
        # - Store dependencies
        pass

    def disable_feature(self, name: str):
        """Disable a feature"""
        # TODO: Disable a feature
        pass

    def enable_feature(self, name: str):
        """Enable a feature"""
        # TODO: Enable a feature
        pass

    def is_enabled(self, name: str) -> bool:
        """Check if feature is enabled"""
        # TODO: Check if feature is enabled
        # - Return False if not found
        pass

    def update_based_on_health(self, service_health: Dict[str, ServiceHealth]):
        """Update feature flags based on service health"""
        # TODO: Update feature flags based on service health
        # - For each feature:
        #   - Check health of all dependency services
        #   - If any dependency is FAILING, disable feature
        #   - If all dependencies are HEALTHY, enable feature
        pass

    def enabled_features(self) -> List[str]:
        """Return list of enabled features"""
        # TODO: Return list of enabled features
        pass


import unittest


class TestGracefulDegradation(unittest.TestCase):
    def test_graceful_degradation_creation(self):
        degradation = GracefulDegradation(1.0, 3)
        self.assertIsNone(degradation.get_health("test"))

    def test_register_and_call_service(self):
        degradation = GracefulDegradation(1.0, 3)

        degradation.register_service("test", lambda: 42)

        result = degradation.call_service("test")
        self.assertEqual(result, 42)
        self.assertEqual(degradation.get_health("test"), ServiceHealth.HEALTHY)

    def test_fallback_on_primary_failure(self):
        degradation = GracefulDegradation(1.0, 3)

        def primary():
            raise Exception("primary failed")

        def fallback():
            return 99

        degradation.register_service("test", primary, fallback)

        result = degradation.call_service("test")
        self.assertEqual(result, 99)  # Fallback succeeded

    def test_cache_on_all_failures(self):
        degradation = GracefulDegradation(1.0, 3)

        # First register with working primary
        degradation.register_service("test", lambda: 42)

        # First call succeeds and caches
        degradation.call_service("test")

        # Update service to fail
        degradation.register_service("test", lambda: (_ for _ in ()).throw(Exception("fail")))

        # Should return cached value
        result = degradation.call_service("test")
        self.assertEqual(result, 42)

    def test_health_degradation(self):
        degradation = GracefulDegradation(1.0, 2)

        def failing():
            raise Exception("fail")

        degradation.register_service("test", failing)

        # First failure
        try:
            degradation.call_service("test")
        except:
            pass
        self.assertEqual(degradation.get_health("test"), ServiceHealth.DEGRADED)

        # Second failure (reaches threshold)
        try:
            degradation.call_service("test")
        except:
            pass
        self.assertEqual(degradation.get_health("test"), ServiceHealth.FAILING)

    def test_quality_levels(self):
        degradation = GracefulDegradation(1.0, 2)

        degradation.register_service("test", lambda: 42)

        # Initially healthy
        degradation.call_service("test")
        self.assertEqual(degradation.get_quality_level("test"), QualityLevel.FULL)

        # Cause degradation
        degradation.register_service("test", lambda: (_ for _ in ()).throw(Exception("fail")))
        try:
            degradation.call_service("test")
        except:
            pass
        self.assertEqual(degradation.get_quality_level("test"), QualityLevel.REDUCED)

    def test_reset_health(self):
        degradation = GracefulDegradation(1.0, 2)

        def failing():
            raise Exception("fail")

        degradation.register_service("test", failing)

        # Cause failures
        for _ in range(2):
            try:
                degradation.call_service("test")
            except:
                pass

        self.assertEqual(degradation.get_health("test"), ServiceHealth.FAILING)

        # Reset
        degradation.reset_health("test")
        self.assertEqual(degradation.get_health("test"), ServiceHealth.HEALTHY)

    def test_feature_flags(self):
        flags = FeatureFlags()

        flags.register_feature("search", ["search-service"])
        self.assertTrue(flags.is_enabled("search"))

        flags.disable_feature("search")
        self.assertFalse(flags.is_enabled("search"))

        flags.enable_feature("search")
        self.assertTrue(flags.is_enabled("search"))

    def test_feature_flags_health_based(self):
        flags = FeatureFlags()

        flags.register_feature("recommendations", ["rec-service"])

        health = {"rec-service": ServiceHealth.FAILING}

        flags.update_based_on_health(health)

        # Feature should be disabled due to failing dependency
        self.assertFalse(flags.is_enabled("recommendations"))

    def test_enabled_features_list(self):
        flags = FeatureFlags()

        flags.register_feature("feature1", [])
        flags.register_feature("feature2", [])
        flags.register_feature("feature3", [])

        flags.disable_feature("feature2")

        enabled = sorted(flags.enabled_features())

        self.assertEqual(enabled, ["feature1", "feature3"])

    def test_manual_cache_update(self):
        degradation = GracefulDegradation(1.0, 3)

        def failing():
            raise Exception("fail")

        degradation.register_service("test", failing)

        degradation.update_cache("test", 123)

        result = degradation.call_service("test")
        self.assertEqual(result, 123)

    def test_clear_cache(self):
        degradation = GracefulDegradation(1.0, 3)

        degradation.register_service("test", lambda: 42)

        # Cache value
        degradation.call_service("test")

        degradation.clear_cache("test")

        # Update to failing
        degradation.register_service("test", lambda: (_ for _ in ()).throw(Exception("fail")))

        # Should fail (no cache)
        with self.assertRaises(Exception):
            degradation.call_service("test")


if __name__ == '__main__':
    unittest.main()

File: Traditional-Important-Algorithms/exercises/09_rate_limiting/rate09_graceful_degradation.rs
// rate09_graceful_degradation.rs
//
// Graceful Degradation is a fault tolerance pattern that maintains partial
// functionality when a system component fails, rather than complete failure.
//
// Strategies:
// - Fallback to cached/default values
// - Reduced functionality mode
// - Alternative service endpoints
// - Feature toggling based on system health
//
// Your task: Implement graceful degradation strategies.
//
// Key concepts:
// - Service health monitoring
// - Fallback mechanisms
// - Quality of service levels
// - Partial failure handling

// I AM NOT DONE

use std::collections::HashMap;
use std::time::{Duration, Instant};

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum ServiceHealth {
    Healthy,
    Degraded,
    Failing,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum QualityLevel {
    Full,       // All features enabled
    Reduced,    // Some features disabled
    Minimal,    // Only critical features
}

pub struct ServiceEndpoint<T> {
    primary: Box<dyn Fn() -> Result<T, String>>,
    fallback: Option<Box<dyn Fn() -> Result<T, String>>>,
    cache: Option<T>,
    health: ServiceHealth,
    last_success: Option<Instant>,
    failure_count: usize,
}

pub struct GracefulDegradation<T> {
    services: HashMap<String, ServiceEndpoint<T>>,
    health_check_interval: Duration,
    failure_threshold: usize,
}

impl<T: Clone> GracefulDegradation<T> {
    pub fn new(health_check_interval: Duration, failure_threshold: usize) -> Self {
        // TODO: Initialize graceful degradation system
        todo!()
    }

    pub fn register_service(
        &mut self,
        name: String,
        primary: Box<dyn Fn() -> Result<T, String>>,
        fallback: Option<Box<dyn Fn() -> Result<T, String>>>,
    ) {
        // TODO: Register a service with primary and optional fallback
        // - Create ServiceEndpoint with given functions
        // - Initialize health as Healthy
        // - Set failure_count to 0
        todo!()
    }

    pub fn call_service(&mut self, name: &str) -> Result<T, DegradationError> {
        // TODO: Call service with graceful degradation
        //
        // Strategy:
        // 1. Get service, return NotFound if doesn't exist
        // 2. Try primary service
        // 3. If success:
        //    - Update cache with result
        //    - Reset failure_count
        //    - Update health to Healthy
        //    - Update last_success
        //    - Return Ok(result)
        // 4. If failure:
        //    - Increment failure_count
        //    - Update health based on failure_count
        //    - Try fallback if available
        //    - If fallback succeeds, return Ok but keep degraded health
        //    - If no fallback or fallback fails, try cache
        //    - If cache exists, return cached value
        //    - Otherwise return Err(Unavailable)
        todo!()
    }

    pub fn get_health(&self, name: &str) -> Option<ServiceHealth> {
        // TODO: Return health status of service
        todo!()
    }

    pub fn get_quality_level(&self, name: &str) -> Option<QualityLevel> {
        // TODO: Determine quality level based on health
        // - Healthy -> Full
        // - Degraded -> Reduced
        // - Failing -> Minimal
        todo!()
    }

    pub fn update_cache(&mut self, name: &str, value: T) {
        // TODO: Manually update cache for a service
        todo!()
    }

    pub fn clear_cache(&mut self, name: &str) {
        // TODO: Clear cache for a service
        todo!()
    }

    pub fn reset_health(&mut self, name: &str) {
        // TODO: Reset service health to Healthy
        // - Set health to Healthy
        // - Reset failure_count to 0
        todo!()
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum DegradationError {
    NotFound,
    Unavailable,
}

// Feature flag manager for graceful degradation
pub struct FeatureFlags {
    flags: HashMap<String, bool>,
    dependencies: HashMap<String, Vec<String>>, // Feature -> required services
}

impl FeatureFlags {
    pub fn new() -> Self {
        // TODO: Initialize feature flags
        todo!()
    }

    pub fn register_feature(&mut self, name: String, dependencies: Vec<String>) {
        // TODO: Register feature with service dependencies
        // - Add to flags as enabled by default
        // - Store dependencies
        todo!()
    }

    pub fn disable_feature(&mut self, name: &str) {
        // TODO: Disable a feature
        todo!()
    }

    pub fn enable_feature(&mut self, name: &str) {
        // TODO: Enable a feature
        todo!()
    }

    pub fn is_enabled(&self, name: &str) -> bool {
        // TODO: Check if feature is enabled
        // - Return false if not found
        todo!()
    }

    pub fn update_based_on_health(&mut self, service_health: &HashMap<String, ServiceHealth>) {
        // TODO: Update feature flags based on service health
        // - For each feature:
        //   - Check health of all dependency services
        //   - If any dependency is Failing, disable feature
        //   - If all dependencies are Healthy, enable feature
        todo!()
    }

    pub fn enabled_features(&self) -> Vec<String> {
        // TODO: Return list of enabled features
        todo!()
    }
}

impl Default for FeatureFlags {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_graceful_degradation_creation() {
        let degradation: GracefulDegradation<i32> = GracefulDegradation::new(
            Duration::from_secs(1),
            3,
        );
        assert_eq!(degradation.get_health("test"), None);
    }

    #[test]
    fn test_register_and_call_service() {
        let mut degradation = GracefulDegradation::new(Duration::from_secs(1), 3);

        degradation.register_service(
            "test".to_string(),
            Box::new(|| Ok(42)),
            None,
        );

        let result = degradation.call_service("test");
        assert_eq!(result, Ok(42));
        assert_eq!(degradation.get_health("test"), Some(ServiceHealth::Healthy));
    }

    #[test]
    fn test_fallback_on_primary_failure() {
        let mut degradation = GracefulDegradation::new(Duration::from_secs(1), 3);

        degradation.register_service(
            "test".to_string(),
            Box::new(|| Err("primary failed".to_string())),
            Some(Box::new(|| Ok(99))),
        );

        let result = degradation.call_service("test");
        assert_eq!(result, Ok(99)); // Fallback succeeded
    }

    #[test]
    fn test_cache_on_all_failures() {
        let mut degradation = GracefulDegradation::new(Duration::from_secs(1), 3);

        degradation.register_service(
            "test".to_string(),
            Box::new(|| Ok(42)),
            None,
        );

        // First call succeeds and caches
        degradation.call_service("test").ok();

        // Update service to fail
        degradation.register_service(
            "test".to_string(),
            Box::new(|| Err("failed".to_string())),
            None,
        );

        // Should return cached value
        let result = degradation.call_service("test");
        assert_eq!(result, Ok(42));
    }

    #[test]
    fn test_health_degradation() {
        let mut degradation = GracefulDegradation::new(Duration::from_secs(1), 2);

        degradation.register_service(
            "test".to_string(),
            Box::new(|| Err("fail".to_string())),
            None,
        );

        // First failure
        degradation.call_service("test").ok();
        assert_eq!(degradation.get_health("test"), Some(ServiceHealth::Degraded));

        // Second failure (reaches threshold)
        degradation.call_service("test").ok();
        assert_eq!(degradation.get_health("test"), Some(ServiceHealth::Failing));
    }

    #[test]
    fn test_quality_levels() {
        let mut degradation = GracefulDegradation::new(Duration::from_secs(1), 2);

        degradation.register_service(
            "test".to_string(),
            Box::new(|| Ok(42)),
            None,
        );

        // Initially healthy
        degradation.call_service("test").ok();
        assert_eq!(degradation.get_quality_level("test"), Some(QualityLevel::Full));

        // Cause degradation
        degradation.register_service(
            "test".to_string(),
            Box::new(|| Err("fail".to_string())),
            None,
        );
        degradation.call_service("test").ok();
        assert_eq!(degradation.get_quality_level("test"), Some(QualityLevel::Reduced));
    }

    #[test]
    fn test_service_not_found() {
        let mut degradation: GracefulDegradation<i32> = GracefulDegradation::new(
            Duration::from_secs(1),
            3,
        );

        let result = degradation.call_service("nonexistent");
        assert_eq!(result, Err(DegradationError::NotFound));
    }

    #[test]
    fn test_reset_health() {
        let mut degradation = GracefulDegradation::new(Duration::from_secs(1), 2);

        degradation.register_service(
            "test".to_string(),
            Box::new(|| Err("fail".to_string())),
            None,
        );

        // Cause failures
        degradation.call_service("test").ok();
        degradation.call_service("test").ok();
        assert_eq!(degradation.get_health("test"), Some(ServiceHealth::Failing));

        // Reset
        degradation.reset_health("test");
        assert_eq!(degradation.get_health("test"), Some(ServiceHealth::Healthy));
    }

    #[test]
    fn test_feature_flags() {
        let mut flags = FeatureFlags::new();

        flags.register_feature("search".to_string(), vec!["search-service".to_string()]);
        assert!(flags.is_enabled("search"));

        flags.disable_feature("search");
        assert!(!flags.is_enabled("search"));

        flags.enable_feature("search");
        assert!(flags.is_enabled("search"));
    }

    #[test]
    fn test_feature_flags_health_based() {
        let mut flags = FeatureFlags::new();

        flags.register_feature(
            "recommendations".to_string(),
            vec!["rec-service".to_string()],
        );

        let mut health = HashMap::new();
        health.insert("rec-service".to_string(), ServiceHealth::Failing);

        flags.update_based_on_health(&health);

        // Feature should be disabled due to failing dependency
        assert!(!flags.is_enabled("recommendations"));
    }

    #[test]
    fn test_enabled_features_list() {
        let mut flags = FeatureFlags::new();

        flags.register_feature("feature1".to_string(), vec![]);
        flags.register_feature("feature2".to_string(), vec![]);
        flags.register_feature("feature3".to_string(), vec![]);

        flags.disable_feature("feature2");

        let mut enabled = flags.enabled_features();
        enabled.sort();

        assert_eq!(enabled, vec!["feature1", "feature3"]);
    }

    #[test]
    fn test_manual_cache_update() {
        let mut degradation = GracefulDegradation::new(Duration::from_secs(1), 3);

        degradation.register_service(
            "test".to_string(),
            Box::new(|| Err("fail".to_string())),
            None,
        );

        degradation.update_cache("test", 123);

        let result = degradation.call_service("test");
        assert_eq!(result, Ok(123));
    }

    #[test]
    fn test_clear_cache() {
        let mut degradation = GracefulDegradation::new(Duration::from_secs(1), 3);

        degradation.register_service(
            "test".to_string(),
            Box::new(|| Ok(42)),
            None,
        );

        // Cache value
        degradation.call_service("test").ok();

        degradation.clear_cache("test");

        // Update to failing
        degradation.register_service(
            "test".to_string(),
            Box::new(|| Err("fail".to_string())),
            None,
        );

        // Should fail (no cache)
        let result = degradation.call_service("test");
        assert_eq!(result, Err(DegradationError::Unavailable));
    }
}
